File: mocks/artifacts/apply-engagement-metrics.mjs
#!/usr/bin/env node

/**
 * Apply realistic engagement metrics to threads, posts, and AI answers
 * Based on formulas from engagement-patterns.md
 */

import { readFileSync, writeFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const mocksDir = join(__dirname, '..');

// Load data
const threads = JSON.parse(readFileSync(join(mocksDir, 'threads.json'), 'utf-8'));
const posts = JSON.parse(readFileSync(join(mocksDir, 'posts.json'), 'utf-8'));
const aiAnswers = JSON.parse(readFileSync(join(mocksDir, 'ai-answers.json'), 'utf-8'));
const courses = JSON.parse(readFileSync(join(mocksDir, 'courses.json'), 'utf-8'));

// Create course lookup
const courseLookup = {};
courses.forEach(c => {
  courseLookup[c.id] = c;
});

// Demo "now" for consistent calculations
const NOW = new Date('2025-10-07T12:00:00Z');

// Utility: random in range
function randomInRange(min, max) {
  return min + Math.random() * (max - min);
}

// Utility: random integer in range
function randomInt(min, max) {
  return Math.floor(randomInRange(min, max + 1));
}

// Course size factor
function getCourseSizeFactor(enrollmentCount) {
  if (enrollmentCount < 35) return 0.8;  // Small
  if (enrollmentCount > 50) return 1.3;  // Large
  return 1.0;  // Medium
}

// Calculate days since creation
function getDaysSince(dateStr) {
  const date = new Date(dateStr);
  const diff = NOW - date;
  return diff / (1000 * 60 * 60 * 24);
}

// Calculate quality factor for a thread
function calculateQualityFactor(thread, threadPosts, aiAnswer) {
  let score = 0;

  // Has AI answer with high confidence
  if (aiAnswer && aiAnswer.confidenceScore >= 85) {
    score += 0.4;
  } else if (aiAnswer) {
    score += 0.3;
  }

  // Has replies
  if (threadPosts.length > 0) {
    score += 0.2;
  }

  // Has endorsed posts
  const hasEndorsed = threadPosts.some(p => p.endorsed);
  if (hasEndorsed) {
    score += 0.3;
  }

  // Has instructor reply
  const instructorIds = ['user-instructor-1', 'user-instructor-2', 'user-instructor-3'];
  const hasInstructorReply = threadPosts.some(p => instructorIds.includes(p.authorId));
  if (hasInstructorReply) {
    score += 0.2;
  }

  // Thread is resolved
  if (thread.status === 'resolved') {
    score += 0.25;
  }

  return 1 + score;  // Range: 1.0 to 2.25
}

// Calculate thread views
function calculateThreadViews(thread, threadPosts, aiAnswer, course) {
  // Base views by status
  let baseViews;
  if (thread.status === 'resolved') {
    baseViews = randomInRange(20, 35);
  } else if (thread.status === 'answered') {
    baseViews = randomInRange(15, 25);
  } else {
    baseViews = randomInRange(8, 15);
  }

  // Age factor
  const days = getDaysSince(thread.createdAt);
  const ageFactor = Math.min(1 + (days / 7) * 0.5, 2.5);

  // Quality factor
  const qualityFactor = calculateQualityFactor(thread, threadPosts, aiAnswer);

  // Course size factor
  const courseSizeFactor = getCourseSizeFactor(course.enrollmentCount);

  // Calculate views
  const calculatedViews = Math.floor(baseViews * ageFactor * qualityFactor * courseSizeFactor);

  // Cap at 200
  return Math.min(calculatedViews, 200);
}

// Calculate post endorsement probability
function calculatePostEndorsementProbability(post, thread, author) {
  let probability = 0;

  // Role bonus
  if (author && author.role === 'instructor') {
    probability += 0.5;
  } else if (author && author.role === 'ta') {
    probability += 0.3;
  }

  // Content quality heuristics
  if (post.content.length > 200) {
    probability += 0.2;
  }

  if (post.content.includes('```')) {
    probability += 0.25;
  }

  if (/step \d/i.test(post.content)) {
    probability += 0.2;
  }

  // Thread quality correlation
  if (thread.status === 'resolved') {
    probability += 0.3;
  }

  // Cap at 0.95
  return Math.min(probability, 0.95);
}

// Calculate AI answer student endorsements
function calculateStudentEndorsements(aiAnswer, views) {
  const confidence = aiAnswer.confidenceScore;

  if (confidence >= 85) {
    // High confidence
    const base = Math.floor((confidence / 100) * (views / 10) * randomInRange(0.3, 0.6));
    return Math.max(0, base);
  } else if (confidence >= 60) {
    // Medium confidence
    const base = Math.floor((confidence / 100) * (views / 20) * randomInRange(0.2, 0.4));
    return Math.max(0, base);
  } else {
    // Low confidence
    return randomInt(0, 1);
  }
}

// Check if instructor should endorse AI answer
function shouldInstructorEndorse(aiAnswer, thread, views) {
  // Must be high confidence
  if (aiAnswer.confidenceScore < 80) return false;

  // Must have quality citations (2+ citations, relevance >= 85)
  const qualityCitations = aiAnswer.citations.filter(c => c.relevance >= 85);
  if (qualityCitations.length < 2) return false;

  // Thread must be at least 1 day old
  const days = getDaysSince(thread.createdAt);
  if (days < 1) return false;

  // Must have some views
  if (views < 20) return false;

  // Probabilistic approval (40% of qualifying answers)
  return Math.random() < 0.4;
}

// Process threads
console.log('Processing threads...');
const processedThreads = threads.map(thread => {
  const threadPosts = posts.filter(p => p.threadId === thread.id);
  const aiAnswer = aiAnswers.find(a => a.id === thread.aiAnswerId);
  const course = courseLookup[thread.courseId];

  const views = calculateThreadViews(thread, threadPosts, aiAnswer, course);

  return {
    ...thread,
    views
  };
});

// Process posts (endorsements)
console.log('Processing posts...');
// Create a simple author lookup (we don't have users.json loaded, so we'll infer roles from IDs)
const processedPosts = posts.map(post => {
  const thread = processedThreads.find(t => t.id === post.threadId);

  // Infer author role from ID
  let author = { role: 'student' };
  if (post.authorId.includes('instructor')) {
    author.role = 'instructor';
  } else if (post.authorId.includes('ta')) {
    author.role = 'ta';
  }

  const probability = calculatePostEndorsementProbability(post, thread, author);
  const endorsed = Math.random() < probability;

  return {
    ...post,
    endorsed
  };
});

// Process AI answers (endorsements)
console.log('Processing AI answers...');
const processedAiAnswers = aiAnswers.map(aiAnswer => {
  const thread = processedThreads.find(t => t.aiAnswerId === aiAnswer.id);
  const views = thread ? thread.views : 20; // Fallback

  const studentEndorsements = calculateStudentEndorsements(aiAnswer, views);
  const instructorEndorsed = shouldInstructorEndorse(aiAnswer, thread, views);
  const instructorEndorsements = instructorEndorsed ? 1 : 0;

  // Total with boost if instructor endorsed
  let totalEndorsements = studentEndorsements + instructorEndorsements;
  if (instructorEndorsed && studentEndorsements > 0) {
    totalEndorsements += Math.floor(studentEndorsements * 0.3);
  }

  return {
    ...aiAnswer,
    studentEndorsements,
    instructorEndorsements,
    totalEndorsements,
    instructorEndorsed
  };
});

// Save processed data
console.log('\nSaving processed data...');
writeFileSync(
  join(__dirname, 'threads-with-metrics.json'),
  JSON.stringify(processedThreads, null, 2)
);

writeFileSync(
  join(__dirname, 'posts-with-endorsements.json'),
  JSON.stringify(processedPosts, null, 2)
);

writeFileSync(
  join(__dirname, 'ai-answers-with-endorsements.json'),
  JSON.stringify(processedAiAnswers, null, 2)
);

// Calculate statistics
console.log('\n=== ENGAGEMENT METRICS SUMMARY ===\n');

// View statistics
const viewStats = {
  min: Math.min(...processedThreads.map(t => t.views)),
  max: Math.max(...processedThreads.map(t => t.views)),
  avg: Math.round(processedThreads.reduce((sum, t) => sum + t.views, 0) / processedThreads.length)
};

console.log('VIEW DISTRIBUTION:');
console.log(`  Min views: ${viewStats.min}`);
console.log(`  Max views: ${viewStats.max}`);
console.log(`  Avg views: ${viewStats.avg}`);

// Views by status
const viewsByStatus = {
  open: processedThreads.filter(t => t.status === 'open').map(t => t.views),
  answered: processedThreads.filter(t => t.status === 'answered').map(t => t.views),
  resolved: processedThreads.filter(t => t.status === 'resolved').map(t => t.views)
};

console.log('\nVIEWS BY STATUS:');
Object.keys(viewsByStatus).forEach(status => {
  const views = viewsByStatus[status];
  if (views.length > 0) {
    const avg = Math.round(views.reduce((a, b) => a + b, 0) / views.length);
    console.log(`  ${status}: avg ${avg} (${views.length} threads)`);
  }
});

// Post endorsement statistics
const endorsedPosts = processedPosts.filter(p => p.endorsed).length;
const endorsementRate = ((endorsedPosts / processedPosts.length) * 100).toFixed(1);

console.log('\nPOST ENDORSEMENTS:');
console.log(`  Total posts: ${processedPosts.length}`);
console.log(`  Endorsed: ${endorsedPosts} (${endorsementRate}%)`);

// Endorsement by role
const postsByRole = {
  instructor: processedPosts.filter(p => p.authorId.includes('instructor')),
  ta: processedPosts.filter(p => p.authorId.includes('ta')),
  student: processedPosts.filter(p => !p.authorId.includes('instructor') && !p.authorId.includes('ta'))
};

console.log('\nENDORSEMENT RATE BY ROLE:');
Object.keys(postsByRole).forEach(role => {
  const rolePosts = postsByRole[role];
  if (rolePosts.length > 0) {
    const endorsed = rolePosts.filter(p => p.endorsed).length;
    const rate = ((endorsed / rolePosts.length) * 100).toFixed(1);
    console.log(`  ${role}: ${endorsed}/${rolePosts.length} (${rate}%)`);
  }
});

// AI answer endorsement statistics
console.log('\nAI ANSWER ENDORSEMENTS:');

const aiByConfidence = {
  high: processedAiAnswers.filter(a => a.confidenceScore >= 85),
  medium: processedAiAnswers.filter(a => a.confidenceScore >= 60 && a.confidenceScore < 85),
  low: processedAiAnswers.filter(a => a.confidenceScore < 60)
};

Object.keys(aiByConfidence).forEach(tier => {
  const answers = aiByConfidence[tier];
  if (answers.length > 0) {
    const avgStudent = Math.round(answers.reduce((sum, a) => sum + a.studentEndorsements, 0) / answers.length);
    const avgTotal = Math.round(answers.reduce((sum, a) => sum + a.totalEndorsements, 0) / answers.length);
    const instructorCount = answers.filter(a => a.instructorEndorsed).length;
    console.log(`  ${tier} confidence (${answers.length} answers):`);
    console.log(`    Avg student endorsements: ${avgStudent}`);
    console.log(`    Avg total endorsements: ${avgTotal}`);
    console.log(`    Instructor endorsed: ${instructorCount}`);
  }
});

// Quality correlation examples
console.log('\nQUALITY CORRELATION EXAMPLES:');

// High-view threads
const topThreads = [...processedThreads].sort((a, b) => b.views - a.views).slice(0, 3);
topThreads.forEach(thread => {
  const aiAnswer = processedAiAnswers.find(a => a.id === thread.aiAnswerId);
  const endorsements = aiAnswer ? aiAnswer.totalEndorsements : 0;
  const confidence = aiAnswer ? aiAnswer.confidenceScore : 0;
  console.log(`  Thread "${thread.title.slice(0, 50)}..."`);
  console.log(`    Views: ${thread.views}, Status: ${thread.status}, AI confidence: ${confidence}, Endorsements: ${endorsements}`);
});

console.log('\n=== FILES SAVED ===');
console.log('  artifacts/threads-with-metrics.json');
console.log('  artifacts/posts-with-endorsements.json');
console.log('  artifacts/ai-answers-with-endorsements.json');
console.log('\nDone!\n');

File: scripts/apply-engagement-metrics.py
#!/usr/bin/env python3
"""
Apply realistic engagement metrics to mock data based on quality and age.
Implements formulas from doccloud/tasks/mock-data-quality/research/engagement-patterns.md
"""

import json
import random
from datetime import datetime
from pathlib import Path

# Set seed for reproducible demo data
random.seed(42)

# Load data
base_path = Path(__file__).parent.parent / 'mocks'
threads = json.load(open(base_path / 'threads.json'))
posts = json.load(open(base_path / 'posts.json'))
ai_answers = json.load(open(base_path / 'ai-answers.json'))
courses = json.load(open(base_path / 'courses.json'))
users = json.load(open(base_path / 'users.json'))

# Create lookups
courses_by_id = {c['id']: c for c in courses}
posts_by_thread = {}
for post in posts:
    thread_id = post['threadId']
    if thread_id not in posts_by_thread:
        posts_by_thread[thread_id] = []
    posts_by_thread[thread_id].append(post)
ai_by_thread = {a['threadId']: a for a in ai_answers}
users_by_id = {u['id']: u for u in users}

# Demo "now" date
NOW = datetime(2025, 10, 7, 12, 0, 0)

def get_days_since_creation(created_at_str):
    created = datetime.fromisoformat(created_at_str.replace('Z', '+00:00'))
    delta = NOW - created.replace(tzinfo=None)
    return delta.days

def get_course_size_factor(enrollment):
    if enrollment < 35:
        return 0.8  # small
    elif enrollment <= 50:
        return 1.0  # medium
    else:
        return 1.3  # large

def get_base_views(status):
    if status == 'resolved':
        return random.randint(20, 35)
    elif status == 'answered':
        return random.randint(15, 25)
    else:  # open
        return random.randint(8, 15)

def calculate_quality_factor(thread, thread_posts, ai_answer):
    score = 0.0

    # Has AI answer
    if thread.get('hasAIAnswer'):
        score += 0.3

    # Has replies
    if len(thread_posts) > 0:
        score += 0.2

    # Has endorsed posts
    if any(p.get('endorsed') for p in thread_posts):
        score += 0.3

    # Has instructor reply
    for post in thread_posts:
        author = users_by_id.get(post['authorId'])
        if author and author['role'] == 'instructor':
            score += 0.2
            break

    # Thread is resolved
    if thread['status'] == 'resolved':
        score += 0.25

    return 1.0 + score

def calculate_thread_views(thread):
    days = get_days_since_creation(thread['createdAt'])
    course = courses_by_id[thread['courseId']]
    thread_posts = posts_by_thread.get(thread['id'], [])
    ai_answer = ai_by_thread.get(thread['id'])

    base_views = get_base_views(thread['status'])
    age_factor = min(1 + (days / 7) * 0.5, 2.5)  # Cap at 2.5x
    quality_factor = calculate_quality_factor(thread, thread_posts, ai_answer)
    course_size_factor = get_course_size_factor(course['enrollmentCount'])

    calculated = int(base_views * age_factor * quality_factor * course_size_factor)
    return min(calculated, 200)  # Cap at 200

def calculate_ai_student_endorsements(ai_answer, thread_views):
    confidence = ai_answer['confidenceScore']

    if confidence >= 85:  # High confidence
        base = (confidence / 100) * (thread_views / 10) * random.uniform(0.3, 0.6)
    elif confidence >= 60:  # Medium confidence
        base = (confidence / 100) * (thread_views / 20) * random.uniform(0.2, 0.4)
    else:  # Low confidence
        base = random.uniform(0, 2)

    return max(0, int(base))

def should_instructor_endorse(ai_answer, thread_views, days_old):
    # Must be high confidence
    if ai_answer['confidenceScore'] < 80:
        return False

    # Must have quality citations
    quality_citations = [c for c in ai_answer['citations'] if c['relevance'] >= 80]
    if len(quality_citations) < 2:
        return False

    # Thread must have some age
    if days_old < 1:
        return False

    # Thread must have views
    if thread_views < 20:
        return False

    # 40% of qualifying answers get instructor endorsement
    return random.random() < 0.4

# Apply metrics to threads
print("Calculating thread views...")
for thread in threads:
    thread['views'] = calculate_thread_views(thread)
    print(f"  {thread['id']}: {thread['views']} views")

# Apply metrics to AI answers
print("\nCalculating AI answer endorsements...")
for ai_answer in ai_answers:
    thread = next(t for t in threads if t['id'] == ai_answer['threadId'])
    days_old = get_days_since_creation(thread['createdAt'])

    student_endorsements = calculate_ai_student_endorsements(ai_answer, thread['views'])
    instructor_endorsements = 1 if should_instructor_endorse(ai_answer, thread['views'], days_old) else 0

    ai_answer['studentEndorsements'] = student_endorsements
    ai_answer['instructorEndorsements'] = instructor_endorsements
    ai_answer['instructorEndorsed'] = instructor_endorsements > 0

    # Boost total if instructor endorsed
    total = student_endorsements + instructor_endorsements
    if instructor_endorsements > 0:
        total += int(student_endorsements * 0.3)

    ai_answer['totalEndorsements'] = total

    print(f"  {ai_answer['id']}: {student_endorsements} student, {instructor_endorsements} instructor, {total} total")

# Save updated data
print("\nSaving updated files...")
json.dump(threads, open(base_path / 'threads.json', 'w'), indent=2)
json.dump(posts, open(base_path / 'posts.json', 'w'), indent=2)
json.dump(ai_answers, open(base_path / 'ai-answers.json', 'w'), indent=2)

print("\n✓ Engagement metrics applied successfully!")
print(f"\nView distribution: {min(t['views'] for t in threads)}-{max(t['views'] for t in threads)} views")
print(f"Average views: {sum(t['views'] for t in threads) / len(threads):.0f}")
print(f"Endorsed posts: {sum(1 for p in posts if p.get('endorsed', False))}/{len(posts)} ({sum(1 for p in posts if p.get('endorsed', False))/len(posts)*100:.1f}%)")
print(f"AI answers with instructor endorsement: {sum(1 for a in ai_answers if a['instructorEndorsed'])}/{len(ai_answers)}")
